File: backend/jobradar/career/slop.py
"""Deterministic AI-slop detection.

Patterns come from shared/slop_patterns.json (adapted from petergyang/no-ai-slop). This
runs as plain regex before and after any model call, which matters for three reasons: it
costs nothing, it is reproducible, and — unlike asking a model "is this slop?" — it cannot
itself hallucinate a verdict.

Used to grade the language_quality criterion of a CV, and as a gate on every LinkedIn note
and email the agent drafts, since a recruiter reading obvious AI output is worse for the
candidate than plain writing.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path

SEVERITY_COST = {"high": 3.0, "medium": 1.5, "low": 0.5}


@dataclass(frozen=True)
class Hit:
    """One slop pattern found in the text."""

    id: str
    label: str
    severity: str
    fix: str
    excerpt: str
    start: int

    def as_dict(self) -> dict:
        return {
            "id": self.id,
            "label": self.label,
            "severity": self.severity,
            "fix": self.fix,
            "excerpt": self.excerpt,
            "start": self.start,
        }


def shared_dir() -> Path:
    """Locate shared/ by walking up from this file to the repo root."""
    here = Path(__file__).resolve()
    for candidate in here.parents:
        target = candidate / "shared" / "slop_patterns.json"
        if target.exists():
            return candidate / "shared"
    raise FileNotFoundError("could not locate shared/ directory")


@lru_cache(maxsize=1)
def _rules() -> tuple[list[dict], list[dict]]:
    data = json.loads((shared_dir() / "slop_patterns.json").read_text())
    return data["patterns"], data.get("structural_checks", [])


@lru_cache(maxsize=1)
def _compiled() -> list[tuple[dict, re.Pattern[str]]]:
    out = []
    for rule in _rules()[0]:
        try:
            out.append((rule, re.compile(rule["regex"], re.I | re.M)))
        except re.error as exc:  # a bad pattern must not break the whole checker
            raise ValueError(f"slop pattern {rule['id']} is invalid: {exc}") from exc
    return out


def find(text: str, *, kind: str = "cv") -> list[Hit]:
    """Return every slop pattern in `text`, plus structural problems for `kind`.

    `kind` is one of "cv", "linkedin_note" or "email" — it selects which structural
    checks apply, since a 300-character limit means nothing for a CV.
    """
    if not text or not text.strip():
        return []

    hits: list[Hit] = []
    for rule, pattern in _compiled():
        for m in pattern.finditer(text):
            excerpt = " ".join(m.group(0).split())[:90]
            hits.append(
                Hit(
                    id=rule["id"],
                    label=rule["label"],
                    severity=rule.get("severity", "medium"),
                    fix=rule.get("fix", ""),
                    excerpt=excerpt,
                    start=m.start(),
                )
            )

    hits.extend(_structural(text, kind))
    hits.sort(key=lambda h: h.start)
    return hits


def _structural(text: str, kind: str) -> list[Hit]:
    hits: list[Hit] = []
    for check in _rules()[1]:
        if check.get("applies_to") != kind:
            continue
        cid = check["id"]

        if cid == "note_too_long" and len(text) > check["threshold_chars"]:
            hits.append(
                Hit(
                    cid,
                    check["label"],
                    "high",
                    check["fix"],
                    f"{len(text)} characters (limit {check['threshold_chars']})",
                    0,
                )
            )

        elif cid == "email_too_long":
            words = len(text.split())
            if words > check["threshold_words"]:
                hits.append(
                    Hit(
                        cid,
                        check["label"],
                        "medium",
                        check["fix"],
                        f"{words} words (aim for under {check['threshold_words']})",
                        0,
                    )
                )

        elif cid == "bullet_too_long":
            for line, offset in _bullets(text):
                if len(line) > check["threshold_chars"]:
                    hits.append(Hit(cid, check["label"], "medium", check["fix"], line[:90], offset))

        elif cid == "repeated_opening_verb":
            verbs: dict[str, int] = {}
            for line, _ in _bullets(text):
                words = re.findall(r"[A-Za-z']+", line)
                if words:
                    verbs[words[0].lower()] = verbs.get(words[0].lower(), 0) + 1
            for verb, count in verbs.items():
                if count >= check["threshold_count"]:
                    hits.append(
                        Hit(
                            cid,
                            check["label"],
                            "low",
                            check["fix"],
                            f'"{verb}" opens {count} bullets',
                            0,
                        )
                    )
    return hits


def _bullets(text: str) -> list[tuple[str, int]]:
    """Lines that look like CV bullets, with their offset in the source."""
    out = []
    offset = 0
    for raw in text.splitlines():
        line = raw.strip()
        if line[:2] in {"• ", "- ", "* ", "– "} or re.match(r"^[•\-\*–]\s", line):
            out.append((re.sub(r"^[•\-\*–]\s*", "", line), offset))
        offset += len(raw) + 1
    return out


def penalty(hits: list[Hit], *, cap: float = 10.0) -> float:
    """Points to deduct from the language_quality criterion, capped at its full weight."""
    total = sum(SEVERITY_COST.get(h.severity, 1.0) for h in hits)
    return min(total, cap)


def language_score(text: str, *, kind: str = "cv", weight: float = 10.0) -> tuple[float, list[Hit]]:
    """Score language quality out of `weight`, deterministically."""
    hits = find(text, kind=kind)
    return round(max(0.0, weight - penalty(hits, cap=weight)), 1), hits


def clean_report(text: str, *, kind: str = "cv") -> dict:
    hits = find(text, kind=kind)
    by_severity: dict[str, int] = {}
    for h in hits:
        by_severity[h.severity] = by_severity.get(h.severity, 0) + 1
    return {
        "clean": not hits,
        "count": len(hits),
        "by_severity": by_severity,
        "hits": [h.as_dict() for h in hits],
    }

File: frontend/src/lib/slop.ts
/**
 * AI-slop detection in the browser.
 *
 * Reads the same shared/slop_patterns.json the Python checker uses, so a draft judged clean
 * here is judged clean at the terminal too. Pure regex — no model call, no cost, no latency,
 * and no chance of the checker itself hallucinating a verdict.
 */
import patternsJson from "@shared/slop_patterns.json";

export interface SlopHit {
  id: string;
  label: string;
  severity: "high" | "medium" | "low";
  fix: string;
  excerpt: string;
  start: number;
}

export type SlopKind = "cv" | "linkedin_note" | "email";

interface PatternRule {
  id: string;
  label: string;
  regex: string;
  severity: string;
  fix: string;
}

interface StructuralRule {
  id: string;
  label: string;
  applies_to: string;
  fix: string;
  threshold_chars?: number;
  threshold_words?: number;
  threshold_count?: number;
}

const SEVERITY_COST: Record<string, number> = { high: 3, medium: 1.5, low: 0.5 };

const rules = patternsJson as unknown as {
  patterns: PatternRule[];
  structural_checks: StructuralRule[];
};

// Compiled once. The `g` flag is required for matchAll, and each pattern gets a fresh
// lastIndex per call because we rebuild the iterator rather than reusing state.
const compiled = rules.patterns.map((rule) => ({
  rule,
  re: new RegExp(rule.regex, "gim"),
}));

function bullets(text: string): { line: string; offset: number }[] {
  const out: { line: string; offset: number }[] = [];
  let offset = 0;
  for (const raw of text.split("\n")) {
    const line = raw.trim();
    if (/^[•\-*–]\s/.test(line)) {
      out.push({ line: line.replace(/^[•\-*–]\s*/, ""), offset });
    }
    offset += raw.length + 1;
  }
  return out;
}

function structural(text: string, kind: SlopKind): SlopHit[] {
  const hits: SlopHit[] = [];
  for (const check of rules.structural_checks) {
    if (check.applies_to !== kind) continue;

    if (check.id === "note_too_long" && check.threshold_chars && text.length > check.threshold_chars) {
      hits.push({
        id: check.id,
        label: check.label,
        severity: "high",
        fix: check.fix,
        excerpt: `${text.length} characters (limit ${check.threshold_chars})`,
        start: 0,
      });
    } else if (check.id === "email_too_long" && check.threshold_words) {
      const words = text.split(/\s+/).filter(Boolean).length;
      if (words > check.threshold_words) {
        hits.push({
          id: check.id,
          label: check.label,
          severity: "medium",
          fix: check.fix,
          excerpt: `${words} words (aim for under ${check.threshold_words})`,
          start: 0,
        });
      }
    } else if (check.id === "bullet_too_long" && check.threshold_chars) {
      for (const b of bullets(text)) {
        if (b.line.length > check.threshold_chars) {
          hits.push({
            id: check.id,
            label: check.label,
            severity: "medium",
            fix: check.fix,
            excerpt: b.line.slice(0, 90),
            start: b.offset,
          });
        }
      }
    } else if (check.id === "repeated_opening_verb" && check.threshold_count) {
      const verbs = new Map<string, number>();
      for (const b of bullets(text)) {
        const first = b.line.match(/[A-Za-z']+/)?.[0]?.toLowerCase();
        if (first) verbs.set(first, (verbs.get(first) ?? 0) + 1);
      }
      for (const [verb, count] of verbs) {
        if (count >= check.threshold_count) {
          hits.push({
            id: check.id,
            label: check.label,
            severity: "low",
            fix: check.fix,
            excerpt: `"${verb}" opens ${count} bullets`,
            start: 0,
          });
        }
      }
    }
  }
  return hits;
}

export function findSlop(text: string, kind: SlopKind = "cv"): SlopHit[] {
  if (!text?.trim()) return [];

  const hits: SlopHit[] = [];
  for (const { rule, re } of compiled) {
    re.lastIndex = 0;
    for (const m of text.matchAll(re)) {
      hits.push({
        id: rule.id,
        label: rule.label,
        severity: (rule.severity as SlopHit["severity"]) ?? "medium",
        fix: rule.fix,
        excerpt: m[0].replace(/\s+/g, " ").slice(0, 90),
        start: m.index ?? 0,
      });
    }
  }
  hits.push(...structural(text, kind));
  hits.sort((a, b) => a.start - b.start);
  return hits;
}

export function slopPenalty(hits: SlopHit[], cap = 10): number {
  const total = hits.reduce((sum, h) => sum + (SEVERITY_COST[h.severity] ?? 1), 0);
  return Math.min(total, cap);
}

export function languageScore(text: string, kind: SlopKind = "cv", weight = 10) {
  const hits = findSlop(text, kind);
  return {
    score: Math.round(Math.max(0, weight - slopPenalty(hits, weight)) * 10) / 10,
    hits,
  };
}
